### PR TITLE
HSL provider for public-transport-enabled

### PR DESCRIPTION
**Pull request by msjoberg**
I implemented HSL provider (Helsinki Region Transport, Finland: https://www.hsl.fi/en). It has already been used in production in the Transportr android client (formerly Liberario).

### CODE REVIEW

File: enabler/src/de/schildbach/pte/HslProvider.java [hunk 1]
```diff
@@ -0,0 +1,833 @@
...
+	private static final int EARLIER_TRIPS_MINIMUM = 3;
+
+	private String apiBase;
+
+	private final XmlPullParserFactory parserFactory;
+
+	public HslProvider(String user, String pass)
+	{
+		super(NetworkId.HSL);
+		apiBase = String.format("http://api.reittiopas.fi/hsl/%s/?user=%s&pass=%s",
```
Comment (schildbach) [inline enabler/src/de/schildbach/pte/HslProvider.java]: Why are username and password part of the API base? My understanding is that they are constant (http://api.reittiopas.fi/hsl/ in this case) and the parameters are appended dynamically. Username and password could be appended in the apiUri() method.

File: enabler/src/de/schildbach/pte/HslProvider.java [hunk 2]
```diff
@@ -0,0 +1,833 @@
...
+	{
+		XmlPullUtil.enter(pp, "coord");
+		String x = xmlValueTag(pp, "x");
+		String y = xmlValueTag(pp, "y");
+		XmlPullUtil.skipExit(pp, "coord");
+		return new Point(Math.round(Float.parseFloat(y) * COORD_MUL),
+				 Math.round(Float.parseFloat(x) * COORD_MUL));
+	}
+
+	private void xmlSkipUntil(final XmlPullParser pp, final String tagName)
```
Comment (schildbach) [inline enabler/src/de/schildbach/pte/HslProvider.java]: This method could be moved to XmlPullUtil.

**Comment by schildbach** [comment]
Thanks for your contribution. Generally this looks good! I added a couple of in-line comments.

**Comment by msjoberg** [comment]
Thanks for the feedback! You're right, I'm not sure why I formed apiBase in that way. I have now added a commit which has a clearer split into constant part and dynamically appended stuff. I also moved skipUtil to XmlPullUtil.

**Comment by grote** [comment]
Hi @schildbach, just want to remind you of this open request ;)

**Comment by schildbach** [comment]
Thanks for the reminder, I'm aware of it. I'm afraid I won't get back to it until next year.

**Comment by grote** [comment]
Another friendly reminder ;)

**Comment by mvsjober** [comment]
Hello @schildbach, sorry for bugging you, but I noticed this pull request is still waiting. Is there anything else that I need to fix before being suitable for merging?

**Comment by msjoberg** [comment]
Hello @schildbach, sorry for bugging you, but I noticed this pull request is still waiting. Is there anything else that I need to fix before being suitable for merging?

**Comment by schildbach** [comment]
Look at this now…

**Comment by schildbach** [comment]
I just merged, everything sqashed with a couple of changes. Thanks for working on this!

I think it needs a bit of work when parsing dates and times. SimpleDateFormat doesn't handle locales and timezones correctly, so the other providers parse into a GregorianCalendar, initialized to the correct time zone.

**Comment by msjoberg** [comment]
Thanks! I can take a look at how the other providers are doing it and submit another pull request when I have done that.

**Comment by grote** [comment]
I just noticed that the provider is returning addresses without name which I is something I haven't seen in any other provider:
* `Location{ADDRESS, 60121451/24734488, name=Braskarna}`
* `Location{ADDRESS, 60121460/24734326}`
* `Location{ADDRESS, 60231543/24658332}`

**Pull request closed without merging**